### PR TITLE
Bottlerocket support

### PR DESCRIPTION
**Pull request by g-gaston**
I still need to test this once we have a new bottlertocket bootstrap image and ova built

### CODE REVIEW

File: pkg/userdata/bottlerocket/node_userdata.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
+package bottlerocket
```
Comment (g-gaston) [inline pkg/userdata/bottlerocket/node_userdata.go:1]: This file is mostly copy-pasta from vignesh changes for kubeadm

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -26,28 +26,3 @@ rules:
   - get
   - patch
   - update
-- apiGroups:
```
Comment (g-gaston) [inline config/rbac/role.yaml]: Why is this changing?

File: pkg/userdata/bottlerocket/template.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/pkg/errors"
+)
+
+const (
+	cloudConfigHeader = `## template: jinja
+#cloud-config
+`
+	// sentinelFileCommand writes a file to /run/cluster-api to signal successful Kubernetes bootstrapping in a way that
+	// works both for Linux and Windows OS.
+	sentinelFileCommand = "echo success > /run/cluster-api/bootstrap-success.complete"
```
Comment (g-gaston) [inline pkg/userdata/bottlerocket/template.go:17]: @mrajashree I need a bit of help to understand what this is for

File: pkg/userdata/input.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+type EtcdadmArgs struct {
+	Version         string
+	ImageRepository string
+}
+
+func (args *EtcdadmArgs) Flags() []string {
+	flags := make([]string, 0, 3)
+	flags = append(flags, "--init-system kubelet")
```
Comment (g-gaston) [inline pkg/userdata/input.go]: This will run etcd on pods when using cloud-init (I haven't tested it)
We can move it a different PR

File: api/v1alpha3/etcdadmconfig_types.go [hunk 5]
```diff
@@ -50,12 +58,16 @@ type EtcdadmConfigSpec struct {
 	// +optional
 	Version string `json:"version,omitempty"`
 
-	// EtcdReleaseURL is an optional field to specify where etcdadm can download etcd from
+	// ImageRepository is an optional field to specify where etcdadm can pull etcd images from
 	// +optional
-	EtcdReleaseURL string `json:"etcdReleaseURL,omitempty"`
```
Comment (mrajashree) [inline api/v1alpha3/etcdadmconfig_types.go:55]: etcdadm uses this as one of the args when running downloading etcd for running it as a systemd service so we shouldn't remove this

Comment (g-gaston) [reply api/v1alpha3/etcdadmconfig_types.go:55]: it's back :)

File: pkg/userdata/bottlerocket/prepare.go [hunk 6]
```diff
@@ -1,14 +1,36 @@
...
+
+func patchCertPaths(input *userdata.BaseUserData) {
+	// hacky. new array bc I need to update the original object and these are not pointers
+	files := make([]v1alpha3.File, 0, len(input.WriteFiles))
+	for _, file := range input.WriteFiles {
+		if filepath.Dir(file.Path) == orgCertsPath {
+			file.Path = filepath.Join(newCertsPath, filepath.Base(file.Path))
+		}
+		files = append(files, file)
+	}
```
Comment (mrajashree) [inline pkg/userdata/bottlerocket/prepare.go:31]: I think we can still update using an index while ranging over input.WriteFiles? this way we won't need new slice
```
	for ind, file := range input.WriteFiles {
		if filepath.Dir(file.Path) == orgCertsPath {
			file.Path = filepath.Join(newCertsPath, filepath.Base(file.Path))
		}
		files[ind] = file
	}
```

File: pkg/userdata/bottlerocket/template.go (hunk 3 above)
Comment (mrajashree) [reply pkg/userdata/bottlerocket/template.go:17]: just noting here that as per our discussion we don't need this for BR :)

File: pkg/userdata/input.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+type EtcdadmArgs struct {
+	Version         string
+	ImageRepository string
+	EtcdReleaseURL  string
+}
+
+func (args *EtcdadmArgs) SystemdFlags() []string {
+	flags := make([]string, 0, 3)
```
Comment (mrajashree) [inline pkg/userdata/input.go:52]: I think this should only declare the slice instead of initializing with a length since we are using append as it is, plus the release-url and version flags are option

File: pkg/userdata/bottlerocket/prepare.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
```
Comment (mrajashree) [inline pkg/userdata/bottlerocket/prepare.go]: I think we can still update using an index while ranging over input.WriteFiles? this way we won't need new slice


```suggestion
	for ind, file := range input.WriteFiles {
		if filepath.Dir(file.Path) == orgCertsPath {
			file.Path = filepath.Join(newCertsPath, filepath.Base(file.Path))
		}
		files[ind] = file
	}
```

File: pkg/userdata/input.go (hunk 7 above)
Comment (g-gaston) [reply pkg/userdata/input.go:52]: Oh this interesting
This code inits the slice with capacity 3 but length 0
I believe that if I don't allocate memory when I declare it, the go runtime will do it during the first append (and it will possibly allocate way more memory
What do you think?
